File: app/src/main/java/com/petnbu/petnbu/feed/comment/CommentsRecyclerViewAdapter.java
package com.petnbu.petnbu.feed.comment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.RoundedBitmapDrawable;
import android.support.v4.graphics.drawable.RoundedBitmapDrawableFactory;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextUtils;
import android.text.format.DateUtils;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.android.gms.common.internal.Preconditions;
import com.petnbu.petnbu.BaseBindingViewHolder;
import com.petnbu.petnbu.GlideApp;
import com.petnbu.petnbu.GlideRequests;
import com.petnbu.petnbu.R;
import com.petnbu.petnbu.model.Photo;
import com.petnbu.petnbu.util.Utils;
import com.petnbu.petnbu.databinding.ViewCommentBinding;
import com.petnbu.petnbu.databinding.ViewLoadingBinding;
import com.petnbu.petnbu.model.CommentUI;
import com.petnbu.petnbu.model.LocalStatus;
import com.petnbu.petnbu.util.ColorUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class CommentsRecyclerViewAdapter extends RecyclerView.Adapter<BaseBindingViewHolder> {

    private final int VIEW_TYPE_COMMENT = 1;
    private final int VIEW_TYPE_LOADING = 2;

    private List<CommentUI> mComments;
    private CommentsViewModel mCommentsViewModel;
    private String mFeedId;
    private GlideRequests mGlideRequests;

    private int mDataVersion;
    private boolean mAddLoadMore;

    public CommentsRecyclerViewAdapter(Context context, List<CommentUI> comments, String feedId,
                                       CommentsViewModel commentsViewModel) {
        Preconditions.checkNotNull(context);
        Preconditions.checkNotNull(commentsViewModel);

        mComments = comments != null ? comments : new ArrayList<>(0);
        mCommentsViewModel = commentsViewModel;
        mFeedId = feedId;
        mGlideRequests = GlideApp.with(context);
    }

    @NonNull
    @Override
    public BaseBindingViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_LOADING) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_loading, parent, false);
            return new ViewLoadingHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_comment, parent, false);
            return new ViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseBindingViewHolder holder, int position) {
        holder.bindData(getItem(position));
    }

    @Override
    public void onBindViewHolder(@NonNull BaseBindingViewHolder holder, int position, @NonNull List<Object> payloads) {
        if(!payloads.isEmpty()) {
            holder.bindData(getItem(position), payloads);
        } else {
            super.onBindViewHolder(holder, position, payloads);
        }
    }

    @Override
    public int getItemCount() {
        return getDataItemCount() + (mAddLoadMore ? 1 : 0);
    }

    private int getDataItemCount() {
        return mComments.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (position < getDataItemCount() && getDataItemCount() > 0) {
            return VIEW_TYPE_COMMENT;
        }
        return VIEW_TYPE_LOADING;
    }

    public void setAddLoadMore(boolean addLoadMore) {
        if (mAddLoadMore != addLoadMore) {
            if (addLoadMore) {
                mAddLoadMore = addLoadMore;
                notifyItemInserted(getLoadingMoreItemPosition());
            } else {
                notifyItemRemoved(getLoadingMoreItemPosition());
                mAddLoadMore = addLoadMore;
            }
        }
    }

    private CommentUI getItem(int position) {
        if (position < 0 || position >= mComments.size())
            return null;
        return mComments.get(position);
    }

    private int getLoadingMoreItemPosition() {
        return mAddLoadMore ? getItemCount() - 1 : RecyclerView.NO_POSITION;
    }

    @SuppressLint("StaticFieldLeak")
    public void setComments(List<CommentUI> comments) {
        mDataVersion++;
        final int startVersion = mDataVersion;
        new AsyncTask<Void, Void, DiffUtil.DiffResult>() {
            @Override
            protected DiffUtil.DiffResult doInBackground(Void... voids) {
                return DiffUtil.calculateDiff(new CommentsDiffCallback(mComments, comments));
            }

            @Override
            protected void onPostExecute(DiffUtil.DiffResult diffResult) {
                if (startVersion != mDataVersion) {
                    // ignore update
                    return;
                }
                mComments = comments;
                diffResult.dispatchUpdatesTo(CommentsRecyclerViewAdapter.this);
            }
        }.execute();
    }

    protected class ViewHolder extends BaseBindingViewHolder<ViewCommentBinding, CommentUI> {

        private CommentUI mComment;
        private final View.OnClickListener mOpenRepliesClickListener = (v) -> {
            if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                CommentUI commentUI = mComments.get(getAdapterPosition());
                if (commentUI.getLocalStatus() != LocalStatus.STATUS_UPLOADING)
                    mCommentsViewModel.openRepliesForComment(commentUI.getId());
            }
        };

        public ViewHolder(View itemView) {
            super(itemView);
            mBinding.tvReply.setOnClickListener(mOpenRepliesClickListener);
            mBinding.tvLatestComment.setOnClickListener(mOpenRepliesClickListener);
            mBinding.tvPreviousReplies.setOnClickListener(mOpenRepliesClickListener);

            mBinding.imgProfile.setOnClickListener(v -> {
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    CommentUI commentUI = mComments.get(getAdapterPosition());
                    if (commentUI.getLocalStatus() != LocalStatus.STATUS_UPLOADING)
                        mCommentsViewModel.openUserProfile(commentUI.getOwner().getUserId());
                }
            });

            mBinding.imgLike.setOnClickListener(view ->{
                if (getAdapterPosition() != RecyclerView.NO_POSITION) {
                    CommentUI commentUI = mComments.get(getAdapterPosition());
                    if (commentUI.getLocalStatus() != LocalStatus.STATUS_UPLOADING)
                        mCommentsViewModel.likeCommentClicked(commentUI.getId());
                }
            });
        }

        @Override
        public void bindData(CommentUI item) {
            mComment = item;
            displayUserInfo();
            displayContent();
            displayInfo();
            displayReplies();
        }

        @Override
        public void bindData(CommentUI item, List<Object> payloads) {
            mComment = item;

            Bundle bundle = (Bundle) payloads.get(0);
            if(bundle.getBoolean("like_status")) {
                displayLikeInfo();
            }
            if(bundle.getBoolean("latest_comment")) {
                displayReplies();
            }
        }

        private void displayUserInfo() {
            String avatarUrl = TextUtils.isEmpty(mComment.getOwner().getAvatar().getThumbnailUrl()) ?
                    mComment.getOwner().getAvatar().getOriginUrl() : mComment.getOwner().getAvatar().getThumbnailUrl();
            mGlideRequests.load(avatarUrl)
                    .centerInside()
                    .into(mBinding.imgProfile);
        }

        private void displayContent() {
            SpannableStringBuilder builder = new SpannableStringBuilder(mComment.getOwner().getName() + "");
            int start = 0;
            builder.setSpan(new StyleSpan(Typeface.BOLD), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.setSpan(new ForegroundColorSpan(Color.BLACK), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.append("  ");

            start = builder.length();
            builder.append(mComment.getContent());
            builder.setSpan(new ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            mBinding.tvContent.setText(builder);

            if (mComment.getPhoto() != null) {
                mBinding.imgPhoto.setVisibility(View.VISIBLE);
                Photo photo = mComment.getPhoto();

                float ratio = photo.getWidth() / (float) photo.getHeight();
                if (photo.getWidth() > photo.getHeight()) {
                    mBinding.imgPhoto.set(1.0f, 1/ratio);
                } else {
                    mBinding.imgPhoto.set(1.0f/2, 1/(2*ratio));
                }
                String photoUrl = TextUtils.isEmpty(mComment.getPhoto().getSmallUrl()) ?
                        mComment.getPhoto().getOriginUrl() : mComment.getPhoto().getSmallUrl();

                mGlideRequests.load(photoUrl)
                        .centerInside()
                        .into(mBinding.imgPhoto);
            } else {
                mBinding.imgPhoto.setVisibility(View.GONE);
            }
        }

        private void displayInfo() {
            mBinding.tvDate.setText(DateUtils.getRelativeTimeSpanString(mComment.getTimeCreated().getTime(),
                    Calendar.getInstance().getTimeInMillis(), 0L, DateUtils.FORMAT_ABBREV_RELATIVE));

            if (!mComment.getId().equals(mFeedId)) {
                mBinding.divider.setVisibility(View.INVISIBLE);
                mBinding.tvLikesCount.setVisibility(View.VISIBLE);
                mBinding.tvReply.setVisibility(View.VISIBLE);
                mBinding.imgLike.setVisibility(View.VISIBLE);
                displayLikeInfo();
            } else {
                mBinding.tvLikesCount.setVisibility(View.GONE);
                mBinding.tvReply.setVisibility(View.GONE);
                mBinding.imgLike.setVisibility(View.GONE);
                mBinding.divider.setVisibility(View.VISIBLE);
                mBinding.progressBar.setVisibility(View.GONE);
            }
        }

        private void displayLikeInfo() {
            if (mComment.getLocalStatus() == LocalStatus.STATUS_UPLOADING || mComment.isLikeInProgress()) {
                mBinding.imgLike.setVisibility(View.INVISIBLE);
                mBinding.progressBar.setVisibility(View.VISIBLE);
            } else {
                mBinding.imgLike.setVisibility(View.VISIBLE);
                mBinding.progressBar.setVisibility(View.GONE);
                if(mComment.isLiked()){
                    mBinding.imgLike.setImageResource(R.drawable.ic_favorite_red_24dp);
                }else{
                    mBinding.imgLike.setImageResource(R.drawable.ic_favorite_border_black_24dp);
                }
            }
            if (mComment.getLikeCount() > 0) {
                mBinding.tvLikesCount.setVisibility(View.VISIBLE);
                mBinding.tvLikesCount.setText(String.format("%d %s", mComment.getLikeCount(), mComment.getLikeCount() > 1 ? "likes" : "like"));
            } else {
                mBinding.tvLikesCount.setVisibility(View.GONE);
            }
        }

        private void displayReplies() {
            if (mComment.getLatestCommentId() != null) {
                mBinding.imgLatestCommentOwnerProfile.setVisibility(View.VISIBLE);
                mBinding.tvLatestComment.setVisibility(View.VISIBLE);
                mBinding.tvPreviousReplies.setVisibility(View.VISIBLE);

                Context context = itemView.getContext();
                SpannableStringBuilder builder = new SpannableStringBuilder(mComment.getLatestCommentOwnerName() + "");
                int start = 0;
                builder.setSpan(new StyleSpan(Typeface.BOLD), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                builder.setSpan(new ForegroundColorSpan(Color.BLACK), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                builder.append("  ");

                start = builder.length();
                builder.append(mComment.getLatestCommentPhoto() != null ? "replied" : mComment.getLatestCommentContent());
                builder.setSpan(new ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, builder.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                mBinding.tvLatestComment.setText(builder);

                if (mComment.getCommentCount() > 1) {
                    mBinding.tvPreviousReplies.setVisibility(View.VISIBLE);
                    mBinding.tvPreviousReplies.setText(context.getString(R.string.feed_view_previous_replies, mComment.getCommentCount() - 1));
                } else {
                    mBinding.tvPreviousReplies.setVisibility(View.GONE);
                }

                String latestCommentOwnerAvatarUrl = TextUtils.isEmpty(mComment.getLatestCommentOwnerAvatar().getThumbnailUrl()) ?
                        mComment.getLatestCommentOwnerAvatar().getOriginUrl() : mComment.getLatestCommentOwnerAvatar().getThumbnailUrl();
                mGlideRequests
                        .load(latestCommentOwnerAvatarUrl)
                        .centerInside()
                        .into(mBinding.imgLatestCommentOwnerProfile);
            } else {
                mBinding.imgLatestCommentOwnerProfile.setVisibility(View.GONE);
                mBinding.tvLatestComment.setVisibility(View.GONE);
                mBinding.tvPreviousReplies.setVisibility(View.GONE);
            }
        }
    }

    protected class ViewLoadingHolder extends BaseBindingViewHolder<ViewLoadingBinding, Void> {

        public ViewLoadingHolder(View itemView) {
            super(itemView);
        }

        @Override
        public void bindData(Void item) {
        }

        @Override
        public void bindData(Void item, List<Object> payloads) {
        }
    }

    private static class CommentsDiffCallback extends DiffUtil.Callback {

        private final List<CommentUI> oldData;
        private final List<CommentUI> newData;

        private CommentsDiffCallback(List<CommentUI> oldData, List<CommentUI> newData) {
            this.oldData = oldData;
            this.newData = newData;
        }

        @Override
        public int getOldListSize() {
            return oldData.size();
        }

        @Override
        public int getNewListSize() {
            return newData.size();
        }

        @Override
        public boolean areItemsTheSame(int oldPos, int newPos) {
            return oldData.get(oldPos).getId().equals(newData.get(newPos).getId());
        }

        @Override
        public boolean areContentsTheSame(int oldPos, int newPos) {
            return oldData.get(oldPos).equals(newData.get(newPos));
        }

        @Nullable
        @Override
        public Object getChangePayload(int oldItemPosition, int newItemPosition) {
            Bundle bundle = new Bundle();
            CommentUI oldComment = oldData.get(oldItemPosition);
            CommentUI newComment = newData.get(newItemPosition);
            if(oldComment.isLikeInProgress() != newComment.isLikeInProgress() || oldComment.isLiked() != newComment.isLiked()) {
                bundle.putBoolean("like_status", true);
                return bundle;
            }
            if(oldComment.getLatestCommentId() != null &&
                    !oldComment.getLatestCommentId().equals(newComment.getLatestCommentId())) {
                bundle.putBoolean("latest_comment", true);
            }
            return !bundle.isEmpty() ? bundle : super.getChangePayload(oldItemPosition, newItemPosition);
        }
    }
}

File: app/src/main/java/com/petnbu/petnbu/feed/comment/CommentsRecyclerViewAdapter.kt
package com.petnbu.petnbu.feed.comment

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.format.DateUtils
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.petnbu.petnbu.BaseBindingViewHolder
import com.petnbu.petnbu.GlideApp
import com.petnbu.petnbu.GlideRequests
import com.petnbu.petnbu.R
import com.petnbu.petnbu.databinding.ViewCommentBinding
import com.petnbu.petnbu.databinding.ViewLoadingBinding
import com.petnbu.petnbu.model.CommentUI
import com.petnbu.petnbu.model.LocalStatus
import com.petnbu.petnbu.util.ColorUtils
import java.util.*

class CommentsRecyclerViewAdapter(context: Context,
                                  private val feedId: String,
                                  private val commentsViewModel: CommentsViewModel)
    : ListAdapter<CommentUI, BaseBindingViewHolder<*, *>>(CommentDiffCallback()) {

    private val glideRequests: GlideRequests = GlideApp.with(context)

    var addLoadMore: Boolean = false
        set(value) {
            if (field != value) {
                if (value) {
                    field = value
                    notifyItemInserted(loadingMoreItemPosition)
                } else {
                    notifyItemRemoved(loadingMoreItemPosition)
                    field = value
                }
            }
        }

    private val dataItemCount: Int
        get() = super.getItemCount()

    private val loadingMoreItemPosition: Int
        get() = if (addLoadMore) itemCount - 1 else RecyclerView.NO_POSITION

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseBindingViewHolder<*, *> {
        return if (viewType == VIEW_TYPE_LOADING) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.view_loading, parent, false)
            ViewLoadingHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.view_comment, parent, false)
            ViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: BaseBindingViewHolder<*, *>, position: Int) {
        (holder as? ViewHolder)?.bindData(getItem(position))
    }

    override fun onBindViewHolder(holder: BaseBindingViewHolder<*, *>, position: Int, payloads: List<Any>) {
        if (payloads.isNotEmpty())
            (holder as? ViewHolder)?.bindData(getItem(position), payloads)
        else
            super.onBindViewHolder(holder, position, payloads)

    }

    override fun getItemCount() = dataItemCount + if (addLoadMore) 1 else 0

    override fun getItemViewType(position: Int) = if (position < dataItemCount && dataItemCount > 0) VIEW_TYPE_COMMENT else VIEW_TYPE_LOADING

    inner class ViewHolder(itemView: View) : BaseBindingViewHolder<ViewCommentBinding, CommentUI>(itemView) {

        private lateinit var comment: CommentUI
        private val openRepliesClickListener = { _: View ->
            if (adapterPosition != RecyclerView.NO_POSITION) {
                getItem(adapterPosition).run {
                    if (localStatus != LocalStatus.STATUS_UPLOADING)
                        commentsViewModel.openRepliesForComment(id)
                }
            }
        }

        init {
            mBinding.tvReply.setOnClickListener(openRepliesClickListener)
            mBinding.tvLatestComment.setOnClickListener(openRepliesClickListener)
            mBinding.tvPreviousReplies.setOnClickListener(openRepliesClickListener)
            mBinding.imgProfile.setOnClickListener { _: View ->
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    getItem(adapterPosition).run {
                        owner?.run {
                            if (localStatus != LocalStatus.STATUS_UPLOADING)
                                commentsViewModel.openUserProfile(userId)
                        }
                    }
                }
            }

            mBinding.imgLike.setOnClickListener { _: View ->
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    getItem(adapterPosition).run {
                        if (localStatus != LocalStatus.STATUS_UPLOADING)
                            commentsViewModel.likeCommentClicked(id)
                    }
                }
            }
        }

        override fun bindData(item: CommentUI) {
            comment = item
            displayUserInfo()
            displayContent()
            displayInfo()
            displayReplies()
        }

        override fun bindData(item: CommentUI, payloads: List<Any>) {
            comment = item

            (payloads[0] as? Bundle)?.run {
                if(getBoolean("like_status"))
                    displayLikeInfo()
                if(getBoolean("latest_comment"))
                    displayReplies()
            }
        }

        private fun displayUserInfo() {
            comment.owner?.run {
                val avatarUrl = if (!avatar.thumbnailUrl.isNullOrEmpty()) avatar.thumbnailUrl else avatar.originUrl
                glideRequests.load(avatarUrl)
                        .centerInside()
                        .into(mBinding.imgProfile)
            }
        }

        private fun displayContent() {
            comment.owner?.run {
                val contentBuilder = SpannableStringBuilder()
                contentBuilder.apply {
                    var start = 0
                    append(name)
                    setSpan(StyleSpan(Typeface.BOLD), start, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    setSpan(ForegroundColorSpan(Color.BLACK), start, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    append("  ")

                    start = length
                    append(comment.content)
                    setSpan(ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
                mBinding.tvContent.text = contentBuilder
            }

            comment.photo?.run {
                mBinding.imgPhoto.visibility = View.VISIBLE

                val ratio = width / height.toFloat()
                if (width > height) {
                    mBinding.imgPhoto.set(1.0f, 1 / ratio)
                } else {
                    mBinding.imgPhoto.set(1.0f / 2, 1 / (2 * ratio))
                }

                val photoUrl = if (!smallUrl.isNullOrEmpty()) smallUrl else originUrl
                glideRequests.load(photoUrl)
                        .centerInside()
                        .into(mBinding.imgPhoto)
            } ?: kotlin.run {
                mBinding.imgPhoto.visibility = View.GONE
            }
        }

        private fun displayInfo() {
            comment.timeCreated?.run {
                mBinding.tvDate.text = DateUtils.getRelativeTimeSpanString(time,
                        Calendar.getInstance().timeInMillis, 0L, DateUtils.FORMAT_ABBREV_RELATIVE)
            }

            if (comment.id != feedId) {
                mBinding.divider.visibility = View.INVISIBLE
                mBinding.tvLikesCount.visibility = View.VISIBLE
                mBinding.tvReply.visibility = View.VISIBLE
                mBinding.imgLike.visibility = View.VISIBLE
                displayLikeInfo()
            } else {
                mBinding.tvLikesCount.visibility = View.GONE
                mBinding.tvReply.visibility = View.GONE
                mBinding.imgLike.visibility = View.GONE
                mBinding.divider.visibility = View.VISIBLE
                mBinding.progressBar.visibility = View.GONE
            }
        }

        private fun displayLikeInfo() {
            if (comment.localStatus == LocalStatus.STATUS_UPLOADING || comment.likeInProgress) {
                mBinding.imgLike.visibility = View.INVISIBLE
                mBinding.progressBar.visibility = View.VISIBLE
            } else {
                mBinding.imgLike.visibility = View.VISIBLE
                mBinding.progressBar.visibility = View.GONE
                mBinding.imgLike.setImageResource(if (comment.isLiked) R.drawable.ic_favorite_red_24dp else R.drawable.ic_favorite_border_black_24dp)
            }
            if (comment.likeCount > 0) {
                mBinding.tvLikesCount.visibility = View.VISIBLE
                mBinding.tvLikesCount.text = "${comment.likeCount} ${if (comment.likeCount > 1) "likes" else "like"}"
            } else {
                mBinding.tvLikesCount.visibility = View.GONE
            }
        }

        private fun displayReplies() {
            if (comment.latestCommentId != null) {
                mBinding.imgLatestCommentOwnerProfile.visibility = View.VISIBLE
                mBinding.tvLatestComment.visibility = View.VISIBLE
                mBinding.tvPreviousReplies.visibility = View.VISIBLE

                val repliesBuilder = SpannableStringBuilder()
                repliesBuilder.apply {
                    var start = 0
                    append(comment.latestCommentOwnerName)
                    setSpan(StyleSpan(Typeface.BOLD), start, repliesBuilder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    setSpan(ForegroundColorSpan(Color.BLACK), start, repliesBuilder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                    append("  ")

                    start = repliesBuilder.length
                    append(if (comment.latestCommentPhoto != null) "replied" else comment.latestCommentContent)
                    setSpan(ForegroundColorSpan(ColorUtils.modifyAlpha(Color.BLACK, 0.8f)), start, repliesBuilder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
                }
                mBinding.tvLatestComment.text = repliesBuilder

                if (comment.commentCount > 1) {
                    mBinding.tvPreviousReplies.visibility = View.VISIBLE
                    mBinding.tvPreviousReplies.text = itemView.context.getString(R.string.feed_view_previous_replies, comment.commentCount - 1)
                } else {
                    mBinding.tvPreviousReplies.visibility = View.GONE
                }

                comment.latestCommentOwnerAvatar?.run {
                    val latestCommentOwnerAvatarUrl = if (!thumbnailUrl.isNullOrEmpty()) thumbnailUrl else originUrl
                    glideRequests
                            .load(latestCommentOwnerAvatarUrl)
                            .centerInside()
                            .into(mBinding.imgLatestCommentOwnerProfile)
                }
            } else {
                mBinding.imgLatestCommentOwnerProfile.visibility = View.GONE
                mBinding.tvLatestComment.visibility = View.GONE
                mBinding.tvPreviousReplies.visibility = View.GONE
            }
        }
    }

    inner class ViewLoadingHolder(itemView: View) : BaseBindingViewHolder<ViewLoadingBinding, Void>(itemView) {

        override fun bindData(item: Void) {}

        override fun bindData(item: Void, payloads: List<Any>) {}
    }

    private class CommentDiffCallback : DiffUtil.ItemCallback<CommentUI>() {

        override fun areItemsTheSame(oldItem: CommentUI, newItem: CommentUI) = oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: CommentUI, newItem: CommentUI) = oldItem == newItem

        override fun getChangePayload(oldItem: CommentUI, newItem: CommentUI): Any? {
            val bundle = Bundle()
            bundle.apply {
                if (oldItem.likeInProgress != newItem.likeInProgress || oldItem.isLiked != newItem.isLiked) {
                    putBoolean("like_status", true)
                }
                if(oldItem.latestCommentId != newItem.latestCommentId) {
                    putBoolean("latest_comment", true)
                }
            }
            return if(!bundle.isEmpty) bundle else super.getChangePayload(oldItem, newItem)
        }
    }

    companion object {
        private const val VIEW_TYPE_COMMENT = 1
        private const val VIEW_TYPE_LOADING = 2
    }
}
